Gnomad tests 2

tiboloic opened the pull request:
Temporary PR to get comments for future full gnomAD v4 rows extraction
I intend to run it only if the test run is successful

tiboloic committed:
Update gnomadv4_extract_test.py

removed the persist() which is suspected to trigger a cryptic error:
https://batch.hail.populationgenomics.org.au/batches/176611/jobs/2

removed the safe guard after finding out that the maximum size of the rows Hail Table for gnomadv4 should be 20Gb

tiboloic committed:
Merge branch 'main' into gnomad_tests_2

tiboloic committed:
Update dataproc parameters

Issue: running out of HDFS space
secondary workers do not contribute to HDFS: https://cloud.google.com/dataproc/docs/concepts/compute/secondary-vms
* bump up the number of workers to 20
* specify worker disk size
* set the number of secondary workers to 0

tiboloic committed:
rename analysis runner main.py

From now, anal run encapsulating script will be named after the target script with `_main` added to the script name

tiboloic committed:
Add full gnomAD v4 row extraction

tiboloic committed:
Merge branch 'main' into gnomad_tests_2

lgruen commented on gnomad/gnomadv4_extract_rows.py:
```diff
@@ -0,0 +1,30 @@
+
+# keep only the rows and repartition
+# given the drastic size reduction by summarizing the columns, reduce the number of partitions ~50 fold
+ht = mt.rows().repartition(1000, shuffle=True)
```
A shuffle would require preemptible workers, but you don't really need fully balanced partitions here.

```suggestion
ht = mt.rows().naive_coalesce(1000)
```

However, given how problematic repartitioning operations often are, I'd suggest writing out all 50k partitions first and then trying to repartition in a follow-up process, if necessary.

lgruen commented on gnomad/gnomadv4_extract_rows_main.py:
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env python3
+
+"""Extract rows fron gnomAD v4"""
```
```suggestion
"""Extract rows from gnomAD v4"""
```

lgruen commented on gnomad/gnomadv4_extract_rows.py:
```diff
@@ -0,0 +1,30 @@
+hl.init(default_reference='GRCh38')
+
+# gnomAD v4 exomes VDS
+big_vds = hl.vds.read_vds('gs://gnomad/v4.0/raw/exomes/gnomad_v4.0.vds')
+
+# densify as a MatrixTable
+mt = hl.vds.to_dense_mt(big_vds)
```
```suggestion
vds = hl.vds.read_vds('gs://gnomad/v4.0/raw/exomes/gnomad_v4.0.vds')

# densify as a MatrixTable
mt = hl.vds.to_dense_mt(vds)
```

lgruen commented on gnomad/gnomadv4_extract_rows.py:
```diff
@@ -0,0 +1,30 @@
+
+# keep only the rows and repartition
+# given the drastic size reduction by summarizing the columns, reduce the number of partitions ~50 fold
+ht = mt.rows().repartition(1000, shuffle=True)
```
Can you drop any row fields to reduce the size of the table?

tiboloic replied:
Not really, all QC fields are useful

lgruen commented on gnomad/gnomadv4_extract_rows.py:
```diff
@@ -0,0 +1,30 @@
+
+# get alleles frequencies
+mt = mt.transmute_entries(GT=hl.experimental.lgt_to_gt(mt.LGT, mt.LA))
+mt = hl.variant_qc(mt)
```
I'm still not convinced that you shouldn't run some sort of basic variant QC before computing these stats. Otherwise you'll end up counting a lot of noisy, unreliable calls.

tiboloic replied:
No QC needed at this stage for this analysis. Noisy, unreliable calls won't affect significantly the results here

tiboloic committed:
Update gnomad/gnomadv4_extract_rows_main.py

Co-authored-by: Leonhard Gruenschloss <leonhard@gruenschloss.org>

tiboloic committed:
Update gnomad/gnomadv4_extract_rows.py

change `big_vds` for `vds`

Co-authored-by: Leonhard Gruenschloss <leonhard@gruenschloss.org>

tiboloic committed:
Update gnomad/gnomadv4_extract_rows.py

commit Leo's suggestion

Co-authored-by: Leonhard Gruenschloss <leonhard@gruenschloss.org>

tiboloic committed:
Add scripts for 5% cost run

run the extract rows analysis on a subset  of 2500 partitions to estimate cost

tiboloic committed:
Update gnomadv4_cost_run_extract_rows_main.py

add newline at the end

tiboloic commented:
PR is now ready for merge to run cost analysis

lgruen commented on gnomad/gnomadv4_cost_run_extract_rows.py:
```diff
@@ -0,0 +1,37 @@
+# gnomAD v4 exomes VDS
+big_vds = hl.vds.read_vds('gs://gnomad/v4.0/raw/exomes/gnomad_v4.0.vds')
+
+# subset to ~5% of partitions (big_vds has 47960 partitions)
+small_vds = hl.vds.VariantDataset(
+    big_vds.reference_data._filter_partitions(range(2500)),
+    big_vds.variant_data._filter_partitions(range(2500)),
+)
```
Can you make this controlled by a parameter instead of duplicating the full code?

tiboloic replied:
not for a cost run intended to be run only once. Furthermore would be error prone, imagine a typo when passing the number of partitions as a parameters. I rather keep the code for the cost estimation run isolated

tiboloic reviewed (commented):
reviewed

KatalinaBobowik commented on gnomad/gnomadv4_cost_run_extract_rows.py:
```diff
@@ -0,0 +1,37 @@
+hl.init(default_reference='GRCh38')
+
+# gnomAD v4 exomes VDS
+big_vds = hl.vds.read_vds('gs://gnomad/v4.0/raw/exomes/gnomad_v4.0.vds')
```
For better readability, you can define this up at the top, something like VDS = 'gs://gnomad/v4.0/raw/exomes/gnomad_v4.0.vds', then read in the VDS table here

tiboloic replied:
implemented

KatalinaBobowik commented on gnomad/gnomadv4_cost_run_extract_rows.py:
```diff
@@ -0,0 +1,37 @@
+big_vds = hl.vds.read_vds('gs://gnomad/v4.0/raw/exomes/gnomad_v4.0.vds')
+
+# subset to ~5% of partitions (big_vds has 47960 partitions)
+small_vds = hl.vds.VariantDataset(
```
you could also do: 
`n_partitions = big_vds.n_partitions()`, then `big_vds * 0.05`

tiboloic replied:
would need proper checking that the resulting number of partitions is an integer, and since this piece of code is intended to be run once once, it seems superfluous: I would rather  hard code the number of partitions

tiboloic committed:
Update gnomadv4_cost_run_extract_rows.py

set url of gnomADv4 in a variable

KatalinaBobowik reviewed (approved):
apart from small nit picks I mentioned, looks good to me

tiboloic committed:
Update gnomadv4_cost_run_extract_rows.py

change the variable name with gnomadv4 url to CAPS to follow TOBS convention as suggested by Kat
